src/Engine/Component/ParticleComponents/ParticleEffect.cpp

```diff
@@ -0,0 +1,393 @@
+        updateActiveParticles(dt);
+        int attenuations = getAttenuations(dt);
+        float factor = 1.0f;
+        if (attenuations != 0) {
+            factor= pow(m_effectParams->attenuation, attenuations);
+        }
+        for (int i = 0; i < (int)m_activeParticlePositions->size(); i++) {
+            Particle *p = m_particles[m_activeMap[i]];
+            updatePosition(p, dt, factor);
+            m_activeParticlePositions->at(i) = p->position;
+            p->life += dt;
+
+        }
+    }
+    else if (m_effectParams->loop) {
+        if (m_effectParams->rate == 0) {
+            m_life = 0.0f;
+            m_particles = generateParticles();
+            m_activeParticlePositions = getActiveParticlePositions();
+            m_activeParticleOrientationIDs = getActiveParticleOrientationIDs();
+            m_activeMap = getActiveMap();
+            m_nextActivation = getNextActivation();
+            m_nextAttenuation = 0.0f;
+        }
+        else {
+            m_effectParams->effectDuration += m_life;
+        }
+        update(dt);
+    }
+}
+
+
+
+void ParticleEffect::updateActiveParticles(float dt) {
+    
+    if (m_effectParams->rate != 0.0f) {
+        while ((int)m_activeParticlePositions->size() > 0) {
+            Particle *p = m_particles[m_activeMap[0]];
+            if (p->life > m_effectParams->particleDuration) {
+                removeActiveParticle(0);
+            }
+            else {
+                break;
+            }
+        }
+        // Activate new particles based on rate. Cap at n limit and Remove the oldest Particle
+        while (m_life> m_nextActivation) {
+            if (m_activeParticlePositions->size() == m_effectParams->n) {
+                int pid = m_activeMap[0];
+                removeActiveParticle(0);
+                m_particles[pid] = makeParticle(pid);
+                addActiveParticle(pid);
+            }
+            else {
+                int pid;
+                if (m_activeMap.size() == 0) {
+                    pid = 0;
+                }
+                else {
+                    pid = m_activeMap.back() + 1;
+                    if (pid == m_effectParams->n) {
+                        pid = 0;
+                    }
+                }
+                m_particles[pid] = makeParticle(pid);
+                addActiveParticle(pid);
+            }
+            m_nextActivation += (1.0f / m_effectParams->rate);
+        }
+
+    }
+}
+
+void ParticleEffect::updatePosition(Particle *p, float dt, float attenuationFactor) {
+    p->velocity *= attenuationFactor;
+    for (int i = 0; i < (int)m_effectParams->accelerators->size(); i++) {
+        p->velocity += m_effectParams->accelerators->at(i) * dt;
+    }
+    p->position += p->velocity * dt; 
+}
+
+int ParticleEffect::getAttenuations(float dt) {
+    if (m_effectParams->attenuation == 1.0f) {
+        return 0;
+    }
+    else {
+        int attenuations = 0;
+        while (m_life > m_nextAttenuation) {
```

Comment:
Why do we need a while loop here??
Double check my math but this should be the same:
attenuations = ceil((m_life - m_nextAttenuation) / m_attenuationRate)

Reply:
and then m_nextAttenuation = attenuations * m_attenuationRate